### ISSUE TITLE
Unable to init with private repo modules 

### ISSUE THREAD

**Issue by blt** [hashicorp/setup-terraform#33]
I have a terraform project that uses private repo modules. All of the repositories are under my organization's control and we have a CI Github user that is scoped sufficiently to clone these repos. My `terraform init` step is failing with output like:

```
Could not download module "nginx-ingress" (ingress.tf:9) source code from
"git::https://github.com/goodwatercap/terraform-goodwater-nginx-ingress.git":
error downloading
'https://github.com/goodwatercap/terraform-goodwater-nginx-ingress.git':
/usr/bin/git exited with 128: Cloning into
'.terraform/modules/nginx-ingress'...
fatal: could not read Username for 'https://github.com': No such device or
address
```

I'm not sure why this would be. I've instructed the checkout step to persist credentials. Does the specific form of git URL we've got there subvert that somehow? The failing workflow:

```yaml
name: "[prod] terraform plan"

on:
  pull_request:
    branches:
      - prod

jobs:
  Terraform:
    defaults:
      run:
        working-directory: "."
    name: Terraform Plan
    runs-on: ubuntu-latest
    steps:
      - uses: actions/checkout@v2
        with:
          token: ${{ secrets.CI_BOT_TOKEN }}
          persist-credentials: true

      - uses: hashicorp/setup-terraform@v1
        with:
          terraform_version: 0.12.25

      - name: Terraform Init
        id: init
        run: terraform init
```

I don't see in the project's documentation how to set the credentials explicitly. Is there a better way to write the module source URLs or this workflow to accomplish an init under these circumstances?

**Comment by bmurphey** [comment]
@blt Thanks to @alexbde via  https://github.com/hashicorp/terraform-github-actions/issues/120#issuecomment-575093352 and @mataneine via https://github.com/hashicorp/terraform-github-actions/issues/120#issuecomment-598052629, we smashed those two approaches together (note `chmod 600`) to form a working solution.

```
    - name: 'Terraform Init'
      id: init
      working-directory: root
      run: terraform init
      env:
        GIT_SSH_COMMAND: "echo '${{ secrets.GH_SSH_KEY }}' > id_rsa
          && ssh-keyscan github.com > known_hosts
          && chmod 600 id_rsa known_hosts
          && ssh -i ./id_rsa -o UserKnownHostsFile=./known_hosts"
```

Github actions did not appear to let us persist a file beyond the step that created it, so we were forced to go the `env` route.  Hopefully this helps you and anyone else that runs into this issue until there is a better solution.

**Comment by blt** [comment]
Thanks @bmurphey! Makes sense to me.

**Comment by doug-papenthien-by** [comment]
I still think this should be an unresolved bug against the `setup-terraform` GitHub action.  The GH action is supplied to make using terraform GitHub actions easier, but doesn't automatically bootstrap the vendor supplied method for authenticating (`GITHUB_TOKEN`)?  Every other GH action makes use of this concept without forcing the problem onto its users.  I can understand the terraform project itself staying away from this behavior.  But not the GH action.

In any case, after running into this issue myself and hating the idea of scripting my own credential provider with bogus shell scripts or being forced to switch to SSH just for this, I stumbled upon this suggestion:
https://wahlnetwork.com/2020/08/11/using-private-git-repositories-as-terraform-modules/

Basically, you can inject your GITHUB_TOKEN into your git configuration with the following:
`git config --global url."https://oauth2:${{ secrets.GITHUB_TOKEN }}@github.com".insteadOf https://github.com`

Works like a charm.

**Comment by ls-parwinderjit-singh** [comment]
I am facing same issue .. where I can find GH_SSH_KEY ? I am using default github runner

**Comment by franek** [comment]
thanks @doug-papenthien-by for you comment.
In my case, I had to use this syntax : 
> git config --global url."https://oauth2:${{ secrets.MY_TOKEN }}@github.com".insteadOf ssh://git@github.com

I initialize my terraform like this : 
> source               = "git@github.com:org/module"

**Comment by timkye-blueocean-ai** [comment]
@doug-papenthien-by Can you explain where this command goes?

```
git config --global url."https://oauth2:${{ secrets.CI_TOKEN }}@github.com".insteadOf https://github.com
```

I've tried adding it in the same step just before the `terraform deploy`, but that doesn't work (at least with `source = "github.com/org/repo"`.

**Comment by doug-papenthien-by** [comment]
@timkye-blueocean-ai, it should be run before the `terraform init` step as that's when terraform will attempt to download module code from the private repository.

**Comment by timkye-blueocean-ai** [comment]
I was able to get @franek's working with a PAT, but not yours. I have been unable to get a GitHub App token working at all

**Comment by sheeeng** [comment]
https://github.com/hashicorp/setup-terraform/issues/33#issuecomment-673741344 works! But, we will get another warning.

```text
Warning: The `set-output` command is deprecated and will be disabled soon. Please upgrade to using Environment Files. For more information see: https://github.blog/changelog/2022-10-11-github-actions-deprecating-save-state-and-set-output-commands/
```

**Comment by mehmettaskiner** [comment]
@timkye-blueocean-ai I was able to make it work with Github App token using the method described in the docs here: https://docs.github.com/en/developers/apps/building-github-apps/authenticating-with-github-apps#http-based-git-access-by-an-installation

In order to retrieve an app token I used following action: https://github.com/marketplace/actions/action-github-app-token 
below snippet should happen before `terraform init`
```
    - name: get gh token from my-gh-app
      id: gh-app-auth
      uses: getsentry/action-github-app-token@v1
      with:
        app_id: ${{ secrets.GH_APP_ID }}
        private_key: ${{ secrets.GH_APP_PRIVATE_KEY }}
    - name: hack for github internal repo access
      run: git config --global url."https://x-access-token:${{ steps.gh-app-auth.outputs.token }}@github.com".insteadOf ssh://git@github.com
```

my source is defined like: `git@github.com:owner/module.git?ref=someRef`

**Comment by userbradley** [comment]
For what it's worth, I had this issue at work.

We use SSH references for our modules, so we've been able to use deploy keys.

I put together a blog post on this, if anyone is interested

- [Blog Post](https://breadnet.co.uk/terraform-init-on-github-actions-with-private-modules/?mtm_campaign=github&mtm_kwd=terraform-actions-33)

**Comment by acutchin-bitpusher** [comment]
> @blt Thanks to @alexbde via [hashicorp/terraform-github-actions#120 (comment)](https://github.com/hashicorp/terraform-github-actions/issues/120#issuecomment-575093352) and @mataneine via [hashicorp/terraform-github-actions#120 (comment)](https://github.com/hashicorp/terraform-github-actions/issues/120#issuecomment-598052629), we smashed those two approaches together (note `chmod 600`) to form a working solution.
> 
> ```
>     - name: 'Terraform Init'
>       id: init
>       working-directory: root
>       run: terraform init
>       env:
>         GIT_SSH_COMMAND: "echo '${{ secrets.GH_SSH_KEY }}' > id_rsa
>           && ssh-keyscan github.com > known_hosts
>           && chmod 600 id_rsa known_hosts
>           && ssh -i ./id_rsa -o UserKnownHostsFile=./known_hosts"
> ```
> 
> Github actions did not appear to let us persist a file beyond the step that created it, so we were forced to go the `env` route. Hopefully this helps you and anyone else that runs into this issue until there is a better solution.

I'm getting:
